CIRC-734 Publish circulation events required by mod-patron-blocks

alexanderkurash opened the pull request:
Resolves [CIRC-734](https://issues.folio.org/browse/CIRC-734).

Publishes circulation events such as `check out`, `check in`, `declare lost`, `change due date`, `recall request` to PubSub. Check [automated patron blocks design docs](https://wiki.folio.org/display/DD/Automated+Patron+Blocks+feature+design) for more information. 

## Approach
- In order to register mod-circulation as a publisher to PubSub I've added TenantAPI (`POST /_/tenant`) .
- PubSub client has a dependency on the newest version of org.folio.util. I've changed the version from 24 to 29. This is why there's a lot of HTTP status name changes: HTTP_VALIDATION_ERROR -> HTTP_UNPROCESSABLE_ENTITY.

#### TODOS and Open Questions
This shouldn't be merged before mod-patron-blocks is deployed. PubSub doesn't allow to publish events that no one is subscribed to.


## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

marcjohnson-kint commented:
> PubSub client has a dependency on the newest version of org.folio.util. I removed it from circulation's pom.xml (mod-circulation had a dependency on v24). This is why there's a lot of HTTP status name changes: HTTP_VALIDATION_ERROR -> HTTP_UNPROCESSABLE_ENTITY.

If mod-circulation still relies on org.folio.util then I don't think it should be removed from the dependencies, as that would leave it relying on transitive dependencies of the pub-sub client.

Does that make sense?

bohdan-suprun commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -795,6 +795,30 @@
+            "pubsub.subscribers.post"
+          ]
+        },
+        {
+          "methods": [
+            "DELETE"
+          ],
+          "pathPattern": "/_/tenant"
+        }
```
Seems removal is not implemented, can it be removed?

alexanderkurash replied:
Removed

marcjohnson-kint replied:
Should disabling a tenant be supported?

Are publishers registered with pub-sub? If so, should they be unregistered?

alexanderkurash replied:
That's a good question. PubSub readme doesn't mention "unregistration", but there's a DELETE endpoint. It looks from our experience with vagrant box that nothing bad should happen if we don't unregister, but I don't know if it was intended by PubSub module creators.

alexanderkurash replied:
According to @TarasSpashchenko we're supposed to implement DELETE /_/tenant and clean up our registrations. This PR will be updated.

alexanderkurash replied:
Updated - DELETE is handled now

bohdan-suprun commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -795,6 +795,30 @@
           "delay": "1"
         }
       ]
+    },
```
We need a dependency on pub-sub interfaces. Also, please add pub/sub event publish permission to the endpoints' permission lists (i.e. check-in endpoint, check-out, etc.)

alexanderkurash replied:
Added

bohdan-suprun commented on src/main/java/org/folio/circulation/resources/TenantAPI.java:
```diff
@@ -0,0 +1,54 @@
+
+    vertx.executeBlocking(
+      promise -> registerModuleToPubsub(headers, vertx).setHandler(promise::complete),
+      result -> created(new JsonObject()).writeTo(routingContext.response())
```
What will happen if registration is failed?

alexanderkurash replied:
It'll be ignored, which I think is incorrect. I've added a piece that checks for the result and returns server error if registration failed. This way circulation won't be registered in Okapi if it can't register in PubSub.

marcjohnson-kint replied:
Is this acceptable?

As I understand it, this would stop automated patron blocks from working. Is my understanding reasonable?

alexanderkurash replied:
If we're unable to register in PubSub we won't be able to publish messages and they will be lost forever (in current implementation) which can cause some patrons to be incorrectly blocked/unblocked. It's probably not the end of the world right now, but as we start using PubSub for more and more features I think this opens a broader issue of synchronization in messaging systems. One of the possible solutions I can think of is to store events in a temporary storage (e.g. in mod-circulation-storage) and delete them once they're published - this way they won't be lost. 
@marcjohnson-kint I can revert this code to make /_/tenant succeed regardless of PubSub registration and create stories for temporary event storage. What are your thoughts?

marcjohnson-kint replied:
> It's probably not the end of the world right now

Has this been agreed with the PO for the work, that automated patron blocks might silently fail to work under these circumstances?

> but as we start using PubSub for more and more features I think this opens a broader issue of synchronization in messaging systems

Totally agree, that's why I think the transient nature of the event publishing in the current design is a significant challenge.

> One of the possible solutions I can think of is to store events in a temporary storage (e.g. in mod-circulation-storage) and delete them once they're published - this way they won't be lost.

I agree that is a possibility. It is likely only part of the challenge. As an example, if the publisher could not be registered at all when the module is enabled for a tenant, then there would need to be a process for attempting that registration at a later point.

> I can revert this code to make /_/tenant succeed regardless of PubSub registration and create stories for temporary event storage. What are your thoughts?

Personally, my default in these kinds of circumstances is to fail hard and fast. I would likely choose to fail the enabling of the module unless folks were aware of this trade off and it had been agreed that this kind of silent failure is acceptable.

alexanderkurash replied:
@marcjohnson-kint No, it hasn't been agreed with PO. "Fail hard and fast" probably is a better choice then?

BTW, I just noticed that it wasn't actually failing together with PubSub registration. I've changed this piece and added tests to demonstrate both "fail" and "success" scenarios.

marcjohnson-kint replied:
> No, it hasn't been agreed with PO. "Fail hard and fast" probably is a better choice then?

Until FOLIO has some general guidance around this topic that would be my personal preference. There might be circumstances where that is not appropriate, I think those should be discussed with the relevant PO's (and maybe Technical Leads)

alexanderkurash replied:
@TarasSpashchenko also confirmed that until we come up with a more sophisticated approach, fail fast is the way to go.

bohdan-suprun commented on src/main/resources/MessagingDescriptor.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "publications": [
+    {
+      "eventType": "ITEM_CHECK_OUT",
```
Does it make sense to have `CIRCULATION` prefix for all event names, i.e. `CIRCULATION_ITEM_CHECK_OUT`?

This should guaranty name uniqueness.

alexanderkurash replied:
There's no standard for event names right now, but there's a JIRA ticket to develop them - FOLIO-2408. What we agreed on currently is to use past tense (in the comments https://wiki.folio.org/display/DD/Automated+Patron+Blocks+feature+design) and I see that I forgot to change it here, thanks for highlighting this fragment.
We also thought about using module name as prefix, but we're not sure if it makes sense to connect events to a particular module.

bohdan-suprun replied:
Thanks, @alexanderkurash. I'm fine with both naming conventions, assuming that it was agreed.

marcjohnson-kint replied:
I suggested the past tense naming. There are no agreed standards or conventions at the moment :-/

bohdan-suprun commented on src/main/java/org/folio/circulation/services/EventPublishingService.java:
```diff
@@ -0,0 +1,164 @@
+    if (loanAndRelatedRecords.getLoan() != null) {
+      Loan loan = loanAndRelatedRecords.getLoan();
+
+      JsonObject payloadJsonObject = new JsonObject();
```
The event name describes an event of check out an item. It seems to me that payload should contain as less custom data as possible (in this case only `item id`) and subscribers have to fetch all needed additional data themselves.  

Otherwise, we're introducing dependencies on the payload version between all subscribers.

alexanderkurash replied:
I believe one of the reasons for using messaging was to avoid dependencies (including circular dependencies) between modules. If a subscriber needs to call back for more details it would undermine at least part of the idea.
But I agree that we're introducing another dependency. I'm not aware of any plans for using Schema Registry or something similar for message versioning.

bohdan-suprun replied:
Good point, I did not think about circular dependencies. 

Resolving comment.

bohdan-suprun commented on src/main/java/org/folio/circulation/services/EventPublishingService.java:
```diff
@@ -0,0 +1,164 @@
+        if (result) {
+          logger.debug("Event published successfully: {}", event.getId());
+        } else {
+          logger.error("Failed to publish event: {}", event.getId());
```
Should we log the `throwable` as well?

alexanderkurash replied:
Added throwable, event type and payload.

alexanderkurash commented:
> > PubSub client has a dependency on the newest version of org.folio.util. I removed it from circulation's pom.xml (mod-circulation had a dependency on v24). This is why there's a lot of HTTP status name changes: HTTP_VALIDATION_ERROR -> HTTP_UNPROCESSABLE_ENTITY.
> 
> If mod-circulation still relies on org.folio.util then I don't think it should be removed from the dependencies, as that would leave it relying on transitive dependencies of the pub-sub client.
> 
> Does that make sense?

Right. I've changed org.folio.util version in circulation to the version needed by pubsub to avoid transitive dependencies.

marcjohnson-kint commented:
> Right. I've changed org.folio.util version in circulation to the version needed by pubsub to avoid transitive dependencies.

Thank you

marcjohnson-kint commented on src/main/java/org/folio/circulation/services/EventPublishingService.java:
```diff
@@ -0,0 +1,171 @@
+
+  private Clients clients;
+
+  public EventPublishingService(RoutingContext routingContext) {
```
This class seems to have multiple responsibilities, it both constructs the relevant event (sometimes by fetching extra information) and it publishes the event to pub-sub, could these responsibilities be split?

alexanderkurash replied:
Split to EventPublisher and PubSubPublishingService

marcjohnson-kint replied:
What is the difference between those two classes?

alexanderkurash replied:
EventPublisher constructs events payloads to pass to the PubSubPublishingService. It's called from the code that wants to publish an event. PubSubPublishingService is responsible for actual publishing to PubSub. The whole chain looks like "other code -> EventPublisher -> PubSubPublishingService".

marcjohnson-kint replied:
Thanks for explaining that

marcjohnson-kint commented on src/main/java/org/folio/circulation/services/EventPublishingService.java:
```diff
@@ -0,0 +1,171 @@
+      write(payloadJsonObject, LOAN_ID_FIELD, loan.getId());
+      write(payloadJsonObject, DUE_DATE_FIELD, loan.getDueDate());
+
+      publishEvent(ITEM_CHECKED_OUT_EVENT_TYPE, payloadJsonObject.encode());
```
Does this mean that publishing the event happens completely in the background? What happens if publishing fails?

alexanderkurash replied:
It's ignored. Do you think we should fail the whole process (check in, check out etc.)?

marcjohnson-kint replied:
I think we don't have a general answer to that question yet. I think it's potentially a question for the POs.

In the meantime, we've typically been failing the process hard, as that at least makes the circumstances observable.

alexanderkurash replied:
Changed it to fail when can't publish to PubSub

marcjohnson-kint commented on src/main/resources/MessagingDescriptor.json:
```diff
@@ -0,0 +1,28 @@
+      "signed": false
+    },
+    {
+      "eventType": "LOAN_DUE_DATE_UPDATE",
```
Could this be phrased in the past tense?

Does this represent any change in the due date, for any reason, or specifically a manual change to due date?

```suggestion
      "eventType": "LOAN_DUE_DATE_UPDATED",
```

marcjohnson-kint replied:
The associated test uses the language `due date changed` is that a more appropriate domain phrase for this event?

alexanderkurash replied:
Done

marcjohnson-kint commented on src/test/java/api/CirculationTenantAPITests.java:
```diff
@@ -0,0 +1,29 @@
+public class CirculationTenantAPITests extends APITests {
+
+  @Test
+  public void tenantAPIFailsWhenCantRegisterInPubSub() {
```
```suggestion
  public void tenantActivationFailsWhenCannotRegisterWithPubSub() {
```

alexanderkurash replied:
Done

marcjohnson-kint commented on src/test/java/api/CirculationTenantAPITests.java:
```diff
@@ -0,0 +1,29 @@
+  }
+
+  @Test
+  public void tenantAPISucceedsWhenCanRegisterInPubSub() {
```
```suggestion
  public void tenantActivationSucceedsWhenCanRegisterInPubSub() {
```

alexanderkurash replied:
Done

marcjohnson-kint commented on src/test/java/api/CirculationTenantAPITests.java:
```diff
@@ -0,0 +1,29 @@
+  @Test
+  public void tenantAPISucceedsWhenCanRegisterInPubSub() {
+    Response response = tenantAPIFixture.postTenant();
+    assertThat(response.getStatusCode(), is(HttpStatus.HTTP_CREATED.toInt()));
```
How do we know whether the registration has registered the right set of events?

alexanderkurash replied:
We're checking if a correct PubSub endpoint has been called with a valid body (this was added just now).

marcjohnson-kint replied:
Thanks

marcjohnson-kint commented on src/test/java/api/loans/CheckInByBarcodeTests.java:
```diff
@@ -1038,6 +1041,39 @@ public void overdueFineCalculatedCorrectlyWhenHourlyFeeFinePolicyIsApplied() {
+
+    List<JsonObject> publishedEvents = Awaitility.await()
+      .atMost(1, TimeUnit.SECONDS)
+      .until(FakePubSub::getPublishedEvents, hasSize(2));
```
Why are there 2 events published?

alexanderkurash replied:
First event is for check-out, second is for check-in

marcjohnson-kint replied:
Could a comment be added to explain this or the events be cleared prior to the check in?

alexanderkurash replied:
Added

marcjohnson-kint commented on src/test/java/api/loans/CheckInByBarcodeTests.java:
```diff
@@ -1038,6 +1041,39 @@ public void overdueFineCalculatedCorrectlyWhenHourlyFeeFinePolicyIsApplied() {
+
+    JsonObject event = publishedEvents.get(1);
+
+    assertThat(event, isCheckedInEvent());
+    assertThat(new JsonObject(event.getString("eventPayload")),
+      isValidCheckedInEventPayload(checkedInLoan));
```
Could this detail be better hidden by defining a matcher for the event?

alexanderkurash replied:
Done

marcjohnson-kint commented on src/test/java/api/support/fakes/FakePubSub.java:
```diff
@@ -0,0 +1,72 @@
+    return publishedEvents;
+  }
+
+  public static void cleanUp() {
```
As this is quite different to other fake APIs, could this be explicit about what it is doing?

```suggestion
  public static void clearPublishedEvents() {
```

alexanderkurash replied:
Renamed

marcjohnson-kint commented on src/test/java/api/support/matchers/EventMatcher.java:
```diff
@@ -0,0 +1,71 @@
+
+import io.vertx.core.json.JsonObject;
+
+public class EventMatcher {
```
```suggestion
public class EventMatchers {
```

alexanderkurash replied:
Renamed

marcjohnson-kint commented on src/test/java/api/support/matchers/EventMatcher.java:
```diff
@@ -0,0 +1,71 @@
+
+  public static Matcher<JsonObject> isCheckedOutEvent() {
+    return JsonObjectMatcher.allOfPaths(
+      hasJsonPath("eventType", is(EventPublishingService.ITEM_CHECKED_OUT_EVENT_TYPE))
```
I don't think the API tests should depend upon the production code directly. 

What if the name in the constant was incorrect? 

I think this tends to create the opportunity for false positives and reduce awareness of the impact of changes.

alexanderkurash replied:
Fixed

marcjohnson-kint commented on src/test/java/api/support/matchers/EventMatcher.java:
```diff
@@ -0,0 +1,71 @@
+
+  public static Matcher<JsonObject> isValidCheckedInEventPayload(JsonObject loan) {
+    return JsonObjectMatcher.allOfPaths(
+      hasJsonPath("userId", is(loan.getString("userId"))),
+      hasJsonPath("loanId", is(loan.getString("id"))),
+      hasJsonPath("returnDate", is(loan.getString(LoanProperties.RETURN_DATE)))
```
Why do some of these checks use literals and the other uses a constant from the production code?

alexanderkurash replied:
Changed all to literals

alexanderkurash commented on src/main/java/org/folio/circulation/resources/CirculationTenantAPI.java:
```diff
@@ -0,0 +1,63 @@
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+
+public class CirculationTenantAPI {
```
Now it's not only for activation, it also handles DELETE /_/tenant

marcjohnson-kint replied:
I think activation can refer to activation and deactivation. 

Personally, I think that is more specific and descriptive than tenant API.

If you don't like `activation` then an alternative could be to rename it to `TenantResource` as that is consistent with current naming. I think `circulation` is redundant, as it is within the circulation module.

alexanderkurash replied:
Renamed to TenantActivationResource

alexanderkurash commented on src/main/java/org/folio/circulation/resources/CirculationTenantAPI.java:
```diff
@@ -0,0 +1,63 @@
+    routeRegistration.create(this::postTenant);
+  }
+
+  public void postTenant(RoutingContext routingContext) {
```
It handles `POST /_/tenant`, I believe Okapi calls it when the module is enabled for the tenant. Also, this class now handles `DELETE /_/tenant`. Probably they should be named something like `enableModuleForTenant` and `disableModuleForTenant` respectively?

marcjohnson-kint replied:
> Probably they should be named something like `enableModuleForTenant` and `disableModuleForTenant` respectively?

I'm happy with explicit names like that. I'm ambivalent about the distinction between activate or enable.

alexanderkurash replied:
Okapi docs refer to this as "enable module for tenant", so I renamed the methods accordingly. It's still a bit misleading though, because we're not enabling or disabling module here, Okapi does.

marcjohnson-kint replied:
> Okapi docs refer to this as "enable module for tenant", so I renamed the methods accordingly. It's still a bit misleading though, because we're not enabling or disabling module here, Okapi does.

Agreed, it is misleading or confusing, from Okapi's perspective, the module is being enabled for a tenant, whereas from the modules perspective it is being asked to prepare itself for the tenant.

alexanderkurash commented on src/main/java/org/folio/circulation/resources/CirculationTenantAPI.java:
```diff
@@ -0,0 +1,63 @@
+
+    Vertx vertx = routingContext.vertx();
+
+    vertx.executeBlocking(
```
I think if the module starts running before we register in PubSub we can loose some events.

marcjohnson-kint replied:
> I think if the module starts running before we register in PubSub we can loose some events.

Why is `executeBlocking` required to achieve that?

If a module is accessible prior to it being enabled for a tenant, then that could happen anyway. If not, then no access should be possible until activation / enablement finishes which I would think is controlled by the response to the client, usually Okapi.

Either way, I don't follow how internally blocking affects that.

alexanderkurash replied:
I think you're right. Okapi shouldn't call module's endpoints until it receives response from `POST /_/tenant`, so it doesn't matter if execution is blocking. Changed to non-blocking.

marcjohnson-kint replied:
Thanks

alexanderkurash commented on src/main/java/org/folio/circulation/services/EventPublishingService.java:
```diff
@@ -0,0 +1,171 @@
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.web.RoutingContext;
+
+public class EventPublishingService {
```
This class is now responsible not only for publishing, but also for registering and unregistering in PubSub. I've renamed it to `PubSubService`.

marcjohnson-kint replied:
Could these responsibilities be split up rather than adding them into a larger service with more general scope (and vaguer name)?

alexanderkurash replied:
Split to two services - EventPublisher and PubSubRegistrationService

alexanderkurash commented on src/main/java/org/folio/circulation/services/EventPublishingService.java:
```diff
@@ -0,0 +1,171 @@
+
+  public EventPublishingService(RoutingContext routingContext) {
+    okapiHeaders = routingContext.request().headers().entries().stream()
+      .collect(Collectors.toMap(entry -> entry.getKey().toLowerCase(), Map.Entry::getValue));
```
I can't find a list of headers in WebContext, just the ability to get a header by name. In this case I needed to make all of the header names lowercase.

marcjohnson-kint replied:
Could this method be moved to within the `WebContext`?

alexanderkurash replied:
Moved

marcjohnson-kint commented on src/main/java/org/folio/circulation/resources/TenantActivationResource.java:
```diff
@@ -0,0 +1,67 @@
+
+    Vertx vertx = routingContext.vertx();
+
+    Promise<Object> promise = Promise.promise();
+    promise.future().setHandler(ar -> {
+      if (ar.succeeded()) {
+        created(new JsonObject()).writeTo(routingContext.response());
+      }
+      else {
+        ServerErrorResponse.internalError(routingContext.response(),
+          ar.cause().getLocalizedMessage());
+      }
+    });
+    PubSubRegistrationService.registerModule(headers, vertx, promise);
```
I initially found this confusing to read, as it appeared that the promise was being actioned before the registering with pub-sub. 

Could the registration service return a promise or future, that the handler could be associated with, which would make this read more like a sequence?

alexanderkurash replied:
Rewritten

marcjohnson-kint replied:
Thanks

marcjohnson-kint commented on src/main/java/org/folio/circulation/services/PubSubPublishingService.java:
```diff
@@ -0,0 +1,54 @@
+
+    OkapiConnectionParams params = new OkapiConnectionParams(okapiHeaders, vertxContext.owner());
+
+    PubSubClientUtils.sendEventMessage(event, params)
+      .whenComplete((result, throwable) -> {
+        if (Boolean.TRUE.equals(result)) {
+          logger.debug("Event published successfully. ID: {}, type: {}, payload: {}",
+            event.getId(), event.getEventType(), event.getEventPayload());
+        } else {
+          logger.error("Failed to publish event. ID: {}, type: {}, payload: {}", throwable,
+            event.getId(), event.getEventType(), event.getEventPayload());
+        }
+      });
```
As I understand it, this will fail if there are no subscribers.

Could something similar be put in place to what @bohdan-suprun did in mod-feefines to ignore that failure?

alexanderkurash replied:
Added a similar check

marcjohnson-kint replied:
Thanks

marcjohnson-kint commented on src/main/java/org/folio/circulation/support/http/server/JsonHttpResponse.java:
```diff
@@ -33,6 +33,10 @@ public static HttpResponse unprocessableEntity(JsonObject body) {
     return new JsonHttpResponse(422, body, null);
   }
 
+  public static HttpResponse noContent() {
+    return new JsonHttpResponse(204, null, null);
+  }
```
If there is no content, why would this be a JSON response?

alexanderkurash replied:
Fixed

marcjohnson-kint commented on src/main/resources/MessagingDescriptor.json:
```diff
@@ -0,0 +1,28 @@
+  "publications": [
+    {
+      "eventType": "ITEM_CHECKED_OUT",
+      "description": "Item check out event",
```
```suggestion
      "description": "Item was checked out",
```

marcjohnson-kint commented on src/main/resources/MessagingDescriptor.json:
```diff
@@ -0,0 +1,28 @@
+    },
+    {
+      "eventType": "ITEM_CHECKED_IN",
+      "description": "Item check in event",
```
```suggestion
      "description": "Item was checked in",
```

marcjohnson-kint commented on src/main/resources/MessagingDescriptor.json:
```diff
@@ -0,0 +1,28 @@
+    },
+    {
+      "eventType": "ITEM_DECLARED_LOST",
+      "description": "Item declared lost event",
```
```suggestion
      "description": "Item has been declared lost",
```

marcjohnson-kint commented on src/main/resources/MessagingDescriptor.json:
```diff
@@ -0,0 +1,28 @@
+    },
+    {
+      "eventType": "LOAN_DUE_DATE_CHANGED",
+      "description": "Loan due date update event",
```
```suggestion
      "description": "Due date for a loan has been changed",
```

marcjohnson-kint commented on src/test/java/api/TenantActivationResourceTests.java:
```diff
@@ -0,0 +1,52 @@
+    assertThat(response.getStatusCode(), is(HTTP_INTERNAL_SERVER_ERROR.toInt()));
+  }
+
+  @Test
+  public void tenantActivationSucceedsWhenCanRegisterInPubSub() {
+    Response response = tenantAPIFixture.postTenant();
+    assertThat(response.getStatusCode(), is(HTTP_CREATED.toInt()));
+  }
```
How does the test know that the registration has succeeded?

alexanderkurash replied:
We're checking if a correct PubSub endpoint has been called with a valid body (this was added just now).

marcjohnson-kint commented on src/test/java/api/TenantActivationResourceTests.java:
```diff
@@ -0,0 +1,52 @@
+    assertThat(response.getStatusCode(), is(HTTP_INTERNAL_SERVER_ERROR.toInt()));
+  }
+
+  @Test
+  public void tenantDeactivationSucceedsWhenCanUnregisterInPubSub() {
+    Response response = tenantAPIFixture.deleteTenant();
+    assertThat(response.getStatusCode(), is(HTTP_NO_CONTENT.toInt()));
+  }
```
How does the test know that the de-registration succeeded?

alexanderkurash replied:
Added checks for unregistered publishers' event types

marcjohnson-kint commented on src/test/java/api/loans/DeclareLostAPITests.java:
```diff
@@ -385,6 +389,27 @@ public void canDeclareItemLostIfLostPolicyChargeAmountMissing() {
+
+    List<JsonObject> publishedEvents = Awaitility.await()
+      .atMost(1, TimeUnit.SECONDS)
+      .until(FakePubSub::getPublishedEvents, hasSize(2));
```
Similarly to another comment, could a comment be added to explain why there should be two events or that the events are cleared before the item is declared lost.

alexanderkurash replied:
Added

marcjohnson-kint commented on src/test/java/api/loans/scenarios/ChangeDueDateAPITests.java:
```diff
@@ -260,6 +263,25 @@ public void changeDueDateNoticeIsSentWhenPolicyIsDefined() {
+
+    List<JsonObject> publishedEvents = Awaitility.await()
+      .atMost(1, TimeUnit.SECONDS)
+      .until(FakePubSub::getPublishedEvents, hasSize(2));
```
Similarly to another comment, could a comment be added to explain why there should be two events or that the events are cleared before the loan's due date is changed.

alexanderkurash replied:
Added

marcjohnson-kint commented on src/test/java/api/loans/scenarios/ChangeDueDateAPITests.java:
```diff
@@ -260,6 +263,25 @@ public void changeDueDateNoticeIsSentWhenPolicyIsDefined() {
+
+    JsonObject event = publishedEvents.get(1);
+
+    assertThat(event, isValidLoanDueDateChangedEvent(updatedLoan));
```
Should this event also be published in other circumstances where the due date of the loan changes?

alexanderkurash replied:
I only know of two such circumstances - "due date change" and "recall request", the event is published in both cases. Are there others?

marcjohnson-kint replied:
> I only know of two such circumstances - "due date change" and "recall request", the event is published in both cases. Are there others?

A renewal will likely change the due date. That might be considered different and separate.

I can't remember if a request cancellation or move might also change the due date.

alexanderkurash replied:
I tried to track all of such places and as a result added publishing of "due date changed" event to:
- RenewalResource
- LoanCollectionResource.create
- RequestCollectionResource.replace
- RequestCollectionResource.move

marcjohnson-kint replied:
Have the permissions for publishing a message to pub-sub been added to each endpoint where this code is executed?

marcjohnson-kint commented on src/test/java/api/requests/RequestsAPICreationTests.java:
```diff
@@ -2008,6 +2011,49 @@ public void canFetchHundredRequests() {
+
+    List<JsonObject> publishedEvents = Awaitility.await()
+      .atMost(1, TimeUnit.SECONDS)
+      .until(FakePubSub::getPublishedEvents, hasSize(2));
```
Similarly to another comment, could a comment be added to explain why there should be two events or that the events are cleared before the loan's due date is changed.

alexanderkurash replied:
Added

marcjohnson-kint commented on src/test/java/api/support/fixtures/TenantAPIFixture.java:
```diff
@@ -0,0 +1,31 @@
+import api.support.RestAssuredClient;
+import io.vertx.core.json.JsonObject;
+
+public class TenantAPIFixture {
```
I think we changed the production language to use activation, could we do the same here?

```suggestion
public class TenantActivationFixture {
```

alexanderkurash replied:
Renamed

bohdan-suprun commented on src/main/java/org/folio/circulation/resources/DeclareLostResource.java:
```diff
@@ -45,6 +48,7 @@ private void declareLost(RoutingContext routingContext) {
         .thenCompose(r -> r.after(loan -> lostItemFeeService
           .chargeLostItemFees(loan, request, context.getUserId()))))
       .thenCompose(r -> r.after(storeLoanAndItem::updateLoanAndItemInStorage))
+      .thenComposeAsync(r -> r.after(eventPublisher::publishDeclaredLostEvent))
```
I'm not sure how it is important, but in some circumstances loan can be declared lost and immediately closed

alexanderkurash replied:
Do you mean it can be closed as part of this process or the loan closure can be triggered separately?

bohdan-suprun replied:
The loan can be closed as part of the process if no lost fee is charged (see `lostItemFeeService.chargeLostItemFees` above)

alexanderkurash replied:
For our current purpose it is not important - we only care whether the item is lost. But thanks for this information, it may help us in the future.

bohdan-suprun commented on src/main/java/org/folio/circulation/services/PubSubPublishingService.java:
```diff
@@ -0,0 +1,67 @@
+          logger.error("Failed to publish event. ID: {}, type: {}, payload: {}", throwable,
+            event.getId(), event.getEventType(), event.getEventPayload());
+
+          if (throwable.getMessage().toLowerCase().contains(
```
Possible NPE here.

alexanderkurash replied:
Added null checks

bohdan-suprun commented on src/main/java/org/folio/circulation/services/PubSubRegistrationService.java:
```diff
@@ -0,0 +1,69 @@
+        if (throwable == null) {
+          logger.info("Module was successfully registered as publisher/subscriber in mod-pubsub");
+        } else {
+          logger.error("Error during module registration in mod-pubsub", throwable);
```
Won't it suppress an exception if any?

bohdan-suprun commented on src/test/java/api/TenantActivationResourceTests.java:
```diff
@@ -0,0 +1,66 @@
+    assertThat(response.getStatusCode(), is(HTTP_CREATED.toInt()));
+    assertThat(FakePubSub.getCreatedEventTypes().size(), is(4));
+    assertThat(FakePubSub.getRegisteredPublishers().size(), is(1));
+    assertThat(FakePubSub.getRegisteredSubscribers().size(), is(1));
```
There is no subscriber in the messaging descriptor, why do we expect one?

alexanderkurash replied:
PubSub clients makes a call for subscribers anyway. Though I've noticed there's a ticket to change this behavior and it doesn't make sense to rely on it, so this is removed.

bohdan-suprun replied:
Thank you!

bohdan-suprun commented on src/test/java/api/TenantActivationResourceTests.java:
```diff
@@ -0,0 +1,66 @@
+    assertThat(FakePubSub.getRegisteredPublishers().size(), is(1));
+    assertThat(FakePubSub.getRegisteredSubscribers().size(), is(1));
+
+    assertThat(FakePubSub.getCreatedEventTypes().get(0), isItemCheckedOutEventType());
```
Minor: I'd refactor it to something like 
```
assertThat(FakePubSub.getCreatedEventTypes(), hasItems(
  isItemCheckedOutEventType(),
  isItemCheckedInEventType(),
  isItemDeclaredLostEventType(),
  isLoanDueDateChangedEventType()
));
```

alexanderkurash replied:
Refactored

bohdan-suprun commented on src/test/java/api/support/matchers/EventMatchers.java:
```diff
@@ -0,0 +1,125 @@
+  public static Matcher<JsonObject> isValidItemCheckedOutEvent(JsonObject loan) {
+    return JsonObjectMatcher.allOfPaths(
+      hasJsonPath("eventType", is("ITEM_CHECKED_OUT")),
+      hasJsonPath("eventPayload", isValidItemCheckedOutEventPayload(loan))
```
it can be refactored to something like:
```suggestion
      hasJsonPath("eventPayload", allOf(
         hasJsonPath("userId", is(loan.getString("userId"))),
         hasJsonPath("loanId", is(loan.getString("id"))),
         ...
      ))
```

alexanderkurash replied:
Refactored

bohdan-suprun commented on src/test/java/api/support/matchers/EventMatchers.java:
```diff
@@ -0,0 +1,125 @@
+
+  public static Matcher<JsonObject> isValidItemCheckedOutEvent(JsonObject loan) {
+    return JsonObjectMatcher.allOfPaths(
+      hasJsonPath("eventType", is("ITEM_CHECKED_OUT")),
```
Can the `EventTypeMatchers.isItemCheckedOutEventType` be reused?

alexanderkurash replied:
Refactored

alexanderkurash commented on src/main/java/org/folio/circulation/services/EventPublisher.java:
```diff
@@ -0,0 +1,148 @@
+
+  public EventPublisher(RoutingContext routingContext, Clients circulationClients) {
+    this(routingContext);
+    clients = circulationClients;
```
Removed this constructor

marcjohnson-kint commented:
@alexanderkurash 

As I understand it, this is a prerequisite for https://github.com/folio-org/mod-circulation/pull/557 (@bohdan-suprun please correct me if I am wrong).

Do you know when this work is hoped to be merged?

alexanderkurash commented:
> @alexanderkurash
> 
> As I understand it, this is a prerequisite for #557 (@bohdan-suprun please correct me if I am wrong).
> 
> Do you know when this work is hoped to be merged?

@marcjohnson-kint @bohdan-suprun Hopefully, today or tomorrow

marcjohnson-kint commented:
> > @alexanderkurash
> > As I understand it, this is a prerequisite for #557 (@bohdan-suprun please correct me if I am wrong).
> > Do you know when this work is hoped to be merged?
> 
> @marcjohnson-kint @bohdan-suprun Hopefully, today or tomorrow

@alexanderkurash  If it is likely to go past today, could it be possible to merge this pull request as it currently is, and do any additional work in a follow up pull request?

marcjohnson-kint commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -1985,7 +2032,8 @@
         "scheduled-notice-storage.scheduled-notices.item.post",
-        "users.item.get"
+        "users.item.get",
+        "pubsub.publish.post"
```
The reference UI does not currently use this endpoint for manual change due date, it edits the loan directly using a `PUT /circulation/loans/{id}`, that endpoint will need similar logic and permissions

alexanderkurash replied:
Added publishing logic to the `LoanCollectionResource.replace`, a test for it and PubSub permission to `modperms.circulation.loans.item.put`.

marcjohnson-kint commented on src/main/java/org/folio/circulation/services/EventPublisher.java:
```diff
@@ -0,0 +1,143 @@
+  public CompletableFuture<Result<RequestAndRelatedRecords>> publishDueDateChangedEvent(
+    RequestAndRelatedRecords requestAndRelatedRecords, Clients clients) {
+
+    LoanRepository loanRepository = new LoanRepository(clients);
+    loanRepository.findOpenLoanForRequest(requestAndRelatedRecords.getRequest())
+      .thenCompose(r -> r.after(this::publishDueDateChangedEvent));
```
Please make sure that every endpoint that executes this code has the relevant permissions needed for `findOpenLoanForRequest`?

alexanderkurash replied:
Required permission is `circulation-storage.loans.collection.get`. Permission sets for all three endpoints that execute this code include it:
- `modperms.circulation.requests.item.post`
- `modperms.circulation.requests.item.put`
- `modperms.circulation.requests.item.move.post`

marcjohnson-kint commented on src/main/java/org/folio/circulation/support/http/server/WebContext.java:
```diff
@@ -19,9 +21,13 @@
   public WebContext(RoutingContext routingContext) {
     this.routingContext = routingContext;
+
+    headers = routingContext.request().headers().entries().stream()
+      .collect(Collectors.toMap(entry -> entry.getKey().toLowerCase(), Map.Entry::getValue));
```
Why does this need to be pre-generated rather than during the property getter?

alexanderkurash replied:
I'm not sure if headers can be changed via RoutingContext, they probably can't, but just to be safe I removed the field and moved this to the getter.

marcjohnson-kint commented on src/test/java/api/loans/CheckInByBarcodeTests.java:
```diff
@@ -1049,6 +1051,39 @@ public void canCheckInLostAndPaidItem() {
+
+    JsonObject checkedInLoan = checkInResponse.getLoan();
+
+    // There should be two events published - first one for "check out", second one for "check in"
+
+    List<JsonObject> publishedEvents = Awaitility.await()
```
Personally, I would remove the blank line between the comment and the statement to reinforce that the comment is about that statement

alexanderkurash replied:
Removed blank lines in all similar places

bohdan-suprun commented:
> > > @alexanderkurash
> > > As I understand it, this is a prerequisite for #557 (@bohdan-suprun please correct me if I am wrong).
> > > Do you know when this work is hoped to be merged?
> > 
> > 
> > @marcjohnson-kint @bohdan-suprun Hopefully, today or tomorrow
> 
> @alexanderkurash If it is likely to go past today, could it be possible to merge this pull request as it currently is, and do any additional work in a follow up pull request?

@marcjohnson-kint, @alexanderkurash, I'll create a PR that contains only required for #557 changes, if that's fine. No need for a separate PR for the remaining changes.

marcjohnson-kint commented:
> > > > @alexanderkurash
> > > > As I understand it, this is a prerequisite for #557 (@bohdan-suprun please correct me if I am wrong).
> > > > Do you know when this work is hoped to be merged?
> > > 
> > > 
> > > @marcjohnson-kint @bohdan-suprun Hopefully, today or tomorrow
> > 
> > 
> > @alexanderkurash If it is likely to go past today, could it be possible to merge this pull request as it currently is, and do any additional work in a follow up pull request?
> 
> @marcjohnson-kint, @alexanderkurash, I'll create a PR that contains only required for #557 changes, if that's fine. No need for a separate PR for the remaining changes.

If it is easy to split out only the shared changes for both this and #557 then yes, that would be great, thank you.

bohdan-suprun commented:
@alexanderkurash, @marcjohnson-kint, please take a look at https://github.com/folio-org/mod-circulation/pull/562.
